lapis/connection.py

```diff
@@ -0,0 +1,115 @@
+import random
+
+from usim import Scope, time, Pipe
+
+from lapis.storage import Storage
+from lapis.files import RequestedFile
+from lapis.monitor import sampling_required
+
+
+class Connection(object):
+
+    __slots__ = ("storages", "remote_connection")
+
+    def __init__(self, throughput=100):
+        self.storages = dict()
+        self.remote_connection = Pipe(throughput=throughput)
+
+    def add_storage_element(self, storage_element: Storage):
+        """
+        Register storage element in Connetion module clustering storage elements by
+        sitename
+        :param storage_element:
+        :return:
+        """
+        storage_element.remote_connection = self.remote_connection
+        try:
+            self.storages[storage_element.sitename].append(storage_element)
+        except KeyError:
+            self.storages[storage_element.sitename] = [storage_element]
+
+    async def _determine_inputfile_source(
+        self, requested_file: RequestedFile, dronesite: str, job_repr: str
+    ):
+        """
+        Collects NamedTuples containing the amount of data of the requested file
+        cached in a storage element and the storage element for all reachable storage
+        objects on the drone's site. The tuples are sorted by amount of cached data
+        and the storage object where the biggest part of the file is cached is
+        returned. If the file is not cached in any storage object the connection module
+        remote connection is returned.
+        :param requested_file:
+        :param dronesite:
+        :param job_repr:
+        :return:
+        """
+        provided_storages = self.storages.get(dronesite, None)
+        if provided_storages is not None:
+            look_up_list = []
+            for storage in provided_storages:
+                look_up_list.append(storage.look_up_file(requested_file, job_repr))
```

Comment:
Are you sure that it is better to look up the file in the different storages sequentially and not more parallely? 
I thought that XRootDs behaviour should be represented best if the look up is started at about the same time.

Reply:
This change is similar to my comment on the initial PR #53 (which I can't seem to find anymore...).

To get this out of the way first: **Performing the lookup concurrently is fine by itself**, and something we should add later on again. The issue is *how* this happened combined with it not being needed yet.

At the moment, ``look_up_file`` succeeds immediately (both realtime and simulated). Running it concurrently adds no benefit at all, it only complicates the code. In the interest of maintainability, not using concurrency makes it easier to keep working with the code. *That includes adding concurrency back in later on.*

The old how is difficult to understand and thus verify/maintain/extend. ``look_up_file`` did *either* return directly to the caller *or* return via a side-effect to the queue. Consequently, the caller had to support two different call mechanisms as well. Changing anything about that, e.g. a different mechanism such as "get the first X successful results", would need changing multiple places at once.

-------

As a roadmap for the future, it would be better to use a modular concurrency approach here. What you've implemented is the [gather pattern](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather) (kudos for that, by the way) -- which could be split into a regular async function (``look_up_file``) and the concurrency pattern.

```python3
lookup_list = await usim.gather(
    storage.look_up_file(requested_file, job_repr)
    for storage in provided_storages
)
```

Modularising it like this means:
* We can easily switch between a direct call and concurrency pattern. Neither ``look_up_file`` nor ``Connection`` (or any other usages) have to change significantly if we change the call style.
* We can switch to a different concurrency pattern. For example, the [race pattern](https://trio.readthedocs.io/en/stable/reference-core.html#custom-supervisors) would more closely match XRootD's "wait for X results and discard the rest".